### ISSUE TITLE
Use of internal package not allowed

### ISSUE THREAD

**Issue by mholt** [lucas-clemente/quic-go#797]
Hi guys, as always, great work with this library. Our users love it. :+1:

As you may know, we now vendor all of Caddy's dependencies, quic-go included. But to keep the repo as lightweight as possible, we don't vendor test dependencies. Recently our [CI tests started failing](https://travis-ci.org/mholt/caddy/jobs/267655590) with (from the Caddy repo folder):

```
$ go get -t ./...
vendor/github.com/lucas-clemente/quic-go/integrationtests/tools/testserver/server.go:11:2: use of internal package not allowed
```

I imagine this is because it's trying to use an internal folder from an external package.

Do you understand this more than I do? 😅 I'm not sure the best way to go about fixing it. I'm not sure if vendoring the test deps would help but I'd *really* like to avoid that if possible. I already added over 300k lines of code to our repo when I vendored just the compiled code. (We're vendoring mainly towards reproducible builds, not so much for CI's sake; although it would be nice for the tests to run too!)

**Comment by lucas-clemente** [comment]
We recently reorganized our integration tests (#781), so that's probably the reason why things broke. I'm not sure why that causes issues with vendoring though.

I just tried your command and it worked on my machine (:P) running Go 1.9rc2. IIRC there's a change in 1.9 that removed `vendor/` from the expansion of `./...` – maybe that fixes it?

I tried moving integrationtests/tools to integrationtests/internal/tools, but that didn't fix things. We should probably try understand the actual issue…

**Comment by mholt** [comment]
Upgrading the CI script to Go 1.9 indeed fixed the issue, because ./... ignores the vendor folder. (Annoyingly, `golint ./...` does NOT ignore the vendor folder. Oh well.)

That change breaks builds for anyone not using Go 1.9 (it was only released yesterday) - including the Caddy build server. I wonder if the build error is exposing a code/design smell? I don't know much about `internal` folders but how much of an improvement is the change?

**Comment by mholt** [comment]
As I look at this more, I think I realize that this is a result of the non-internal tests calling functions in an internal package. It doesn't fail for you because your tests and your internal package are in the same folder tree. We don't vendor test dependencies, so the 'internal' folder is being imported from the regular GOPATH which is NOT in the same folder tree, and that's against Go's import rules. I am still not sure this is necessary / a good idea for this reason, and I think the pre-Go-1.9 error indicates this, but Go 1.9 covers up the problem anyway, so I will close this. However, think about it, and if it doesn't sit well with you, it's OK to revert the change. :smile:

**Comment by lucas-clemente** [comment]
Thanks for investigating the issue! I think the issue boils down to the fact that the integrationtests folder shouldn't be looked at at all unless you're interested in the tests. But I don't think there's a way to tell that to Go…

Anyway, as long as it works I don't see the need to fix anything 😉

**Comment by swapnai** [comment]
Hi, 

I am using go version go1.10.3 windows/amd64 and running go test .\... gives me the following. Should I use go 1.9? I am very new to Go and Quic. 
packet_packer.go:11:2: use of internal package not allowed
session.go:15:2: use of internal package not allowed
server.go:12:2: use of internal package not allowed
crypto_stream.go:6:2: use of internal package not allowed
client.go:13:2: use of internal package not allowed
buffer_pool.go:6:2: use of internal package not allowed....

**Comment by Zhiyong-Tsang** [comment]
> Hi,
> 
> I am using go version go1.10.3 windows/amd64 and running go test .... gives me the following. Should I use go 1.9? I am very new to Go and Quic. packet_packer.go:11:2: use of internal package not allowed session.go:15:2: use of internal package not allowed server.go:12:2: use of internal package not allowed crypto_stream.go:6:2: use of internal package not allowed client.go:13:2: use of internal package not allowed buffer_pool.go:6:2: use of internal package not allowed....

Hello, have you fixed it? Could you tell me how to fix it? I use Go 1.19.6. Thanks a lot!
